File: Sorting/selection_sort.cpp
/**
 * @Author: hardcodecoder
 * @Date:   03:10:39  Tuesday 24 March 2020
 * @Last modified by:   hardcodecoder
 * @Last modified time: 03:19:27  Tuesday 24 March 2020
 */


 #include<stdio.h>
 #include<stdlib.h>

//performs selection sort on the array
void selectionSort(int *a, int n) {
  int i, j, min;
  for (i = 0; i < n; i++){
    min = i;                       //Consider a[i] as the smallest value
    for (j = i; j < n; j++){       //Iterate over the array from i+1 to n
      if (a[min] > a[j])           //Check for smaller value than a[i]
      min = j;                   //Assign smaller value index to min
    }

    //If min index is not same as i
    //swap the value a[i] and a[min]
    if (min != i){
      int t = a[i];
      a[i] = a[min];
      a[min] = t;
    }
  }
}


//Driver method
int main() {
  int *a;
  int i, n;

  printf("Enter the size of the array : ");
  scanf("%d", &n);

  a=(int*)malloc(sizeof(int)*n);

  printf("\nEnter the data: ");
  for (i = 0; i < n; i++)
    scanf("%d", &a[i]);

  printf("\nUnsorted data: ");
  for (i = 0; i < n; i++)
    printf("%d ", a[i]);

  selectionSort(a, n);

  printf("\nSorted data: ");
  for (i = 0; i < n; i++)
    printf("%d ", a[i]);

  return 0;
}

File: Sorting/Java/SelectionSort.java
/**
 * @Author: hardcodecoder
 * @Date:   03:08:17 Sunday 17 May 2020
 * @Last modified by:   hardcodecoder
 * @Last modified time: 03:28:36 Sunday 17 May 2020
 */


 import java.io.*;
 import java.util.Scanner;

 public class SelectionSort {

   // Main method
   public static void main(String args[]) {
     new SelectionSort().initialize();
   }

   // Helper method to input data
   private void initialize() {
     System.out.println("Enter the number of elements: ");
     Scanner sc = new Scanner(System.in);
     int n = sc.nextInt();

     float[] list = new float[n];

     System.out.println("Enter the numbers in the arrray:");
     for (int i = 0; i < n; i++)
       list[i] = sc.nextFloat();

     selectionSort(list);
   }


   // Implementation of selection sort algorithm
   private void selectionSort(float[] list) {
     for (int i = 0; i < list.length; i++) {
       int min = i;                                 // Consider list[i] as the smallest value
       for (int j = i; j < list.length; j++) {      // Iterate over the list once for each i
         if (list[j] < list[min])                   // Find if there's a smaller value than list[min]
          min = j;
       }

       // If list[i] is not the smallest value
       // swap list[i] and list[min]
       if (min != i) {
         float f = list[i];
         list[i] = list[min];
         list[min] = f;
       }
     }
     print(list);
   }

   // Helper method to print array
   private void print(float[] list) {
     System.out.print("Sorted array: ");
     for (int i = 0; i < list.length; i++)
       System.out.print(String.format("%.2f, ", list[i]));
     System.out.println();
   }
 }
